docs/reference/development-guide.rst

```diff
@@ -0,0 +1,412 @@
+
+The Omnitrace-specific implementation can be found in 
+`source/lib/omnitrace/library/sampling.cpp <https://github.com/ROCm/omnitrace/blob/main/source/lib/omnitrace/library/sampling.cpp>`_.
+Within `sampling.cpp <https://github.com/ROCm/omnitrace/blob/main/source/lib/omnitrace/library/sampling.cpp>`_, 
+there is a bundle of three sampling components:
+
+* `backtrace_timestamp <https://github.com/ROCm/omnitrace/blob/main/source/lib/omnitrace/library/components/backtrace_timestamp.hpp>`_ simply
+  records the wall-clock time of the sample.
+* `backtrace <https://github.com/ROCm/omnitrace/blob/main/source/lib/omnitrace/library/components/backtrace.hpp>`_
+  records the call-stack via libunwind.
+* `backtrace_metrics <https://github.com/ROCm/omnitrace/blob/main/source/lib/omnitrace/library/components/backtrace_metrics.hpp>`_
+  records the sample metrics, such as peak RSS and the hardware counters.
+
+These three components are bundled together in 
+a tuple-like ``struct`` (``tuple<backtrace_timestamp, backtrace, backtrace_metrics>``).
+A buffer of at least 1024 instances of this tuple is mapped using ``mmap`` 
+per-thread. When this buffer is full, 
+the sampler hands the buffer off to its allocator thread and maps a new buffer with ``mmap``
+before taking the next sample. The allocator thread takes this data 
+and either dynamically stores it in memory or writes it to a file depending on the 
+value of ``OMNITRACE_USE_TEMPORARY_FILES``.
+This schema avoids all allocations in the signal handler, allows the data to grow 
+dynamically, avoids potentially slow I/O within the signal handler, and also enables 
+the capability of avoiding I/O altogether.
+The maximum number of samplers handled by each allocator is governed by the 
+``OMNITRACE_SAMPLING_ALLOCATOR_SIZE`` setting (the default is eight). Whenever an allocator 
+has reached its limit,
+a new internal thread is created to handle the new samplers.
+
+Time-window constraint model
+========================================
+
+With the recent introduction of tracing delay and duration, the 
+`constraint namespace <https://github.com/ROCm/omnitrace/blob/main/source/lib/core/constraint.hpp>`_
+was introduced to improve the management of delays and duration limits for 
+data collection. The ``spec`` class accepts a clock identifier, a delay value, a duration value, and an
+integer indicating how many times to repeat the delay and duration cycle. It is therefore 
+possible to perform tasks such as periodically enabling tracing for brief periods
+of time in between long periods without data collection while the application runs. The
+syntax follows the format ``clock_identifier:delay:capture_duration:cycles``, so a value of 
+``10:1:3`` represents the following sequence of operations:
+
+* 10 seconds where no data is collected, then one second where it is
+* 10 seconds where no data is collected, then one second where it is 
+* 10 seconds where no data is collected, then one second where it is 
```

Comment:
```suggestion
* 10 seconds where no data is collected, then 1 second where it is
* 10 seconds where no data is collected, then 1 second where it is 
* 10 seconds where no data is collected, then 1 second where it is 
```
Sorry, I'm not sure if numerals or words are preferred, I just changed them to be consistent.